src/java/org/apache/cassandra/locator/Replica.java

```diff
@@ -0,0 +1,221 @@
+
+    public final InetAddressAndPort getEndpoint()
+    {
+        return endpoint;
+    }
+
+    public Range<Token> getRange()
+    {
+        return range;
+    }
+
+    public boolean isFull()
+    {
+        return full;
+    }
+
+    public final boolean isTransient()
+    {
+        return !isFull();
+    }
+
+    public ReplicaSet subtract(Replica that)
+    {
+        assert isFull() && that.isFull();  // FIXME: this
+        Set<Range<Token>> ranges = range.subtract(that.range);
+        ReplicaSet replicatedRanges = new ReplicaSet(ranges.size());
+        for (Range<Token> range: ranges)
+        {
+            replicatedRanges.add(new Replica(getEndpoint(), range, isFull()));
+        }
+        return replicatedRanges;
+    }
+
+    /**
+     * Subtract the ranges of the given replicas from the range of this replica,
+     * returning a set of replicas with the endpoint and transient information of
+     * this replica, and the ranges resulting from the subtraction.
+     */
+    public ReplicaSet subtractByRange(Replicas toSubtract)
+    {
+        if (isFull() && Iterables.all(toSubtract, Replica::isFull))
+        {
+            Set<Range<Token>> subtractedRanges = getRange().subtractAll(toSubtract.asRangeSet());
+            ReplicaSet replicaSet = new ReplicaSet(subtractedRanges.size());
+            for (Range<Token> range: subtractedRanges)
+            {
+                replicaSet.add(new Replica(getEndpoint(), range, isFull()));
+            }
+            return replicaSet;
+        }
+        else
+        {
+            // FIXME: add support for transient replicas
+            throw new UnsupportedOperationException("transient replicas are currently unsupported");
+        }
+    }
+
+    public ReplicaList normalizeByRange()
+    {
+        List<Range<Token>> normalized = Range.normalize(Collections.singleton(getRange()));
+        ReplicaList replicas = new ReplicaList(normalized.size());
+        for (Range<Token> normalizedRange: normalized)
+        {
+            replicas.add(new Replica(getEndpoint(), normalizedRange, isFull()));
+        }
+        return replicas;
+    }
+
+    public boolean contains(Range<Token> that)
+    {
+        return getRange().contains(that);
+    }
+
+    public boolean intersectsOnRange(Replica replica)
+    {
+        return getRange().intersects(replica.getRange());
+    }
+
+    public Replica decorateSubrange(Range<Token> subrange)
+    {
+        Preconditions.checkArgument(range.contains(subrange));
+        return new Replica(getEndpoint(), subrange, isFull());
+    }
+
+    public static Replica full(InetAddressAndPort endpoint, Range<Token> range)
+    {
+        return new Replica(endpoint, range, true);
+    }
+
+    /**
+     * We need to assume an endpoint is a full replica in a with unknown ranges in a
+     * few cases, so this returns one that throw an exception if you try to get it's range
+     */
+    public static Replica fullStandin(InetAddressAndPort endpoint)
```

Comment:
I agree it sucks. The problem is that there batchlog and hint related functions that don't work particularly well with the Replica idea, but use the snitch and some write path stuff that were converted to use Replicas to prevent a lot of conversion/copying on the hot path. So using fake replicas solved that problem, but not that well. Let me take another look at the places that use stand ins though, there's probably something less crappy that can be done.